ToolBox Overhaul

jmchilton opened the pull request:
- Implement a tool cache that allows for rapid toolbox reloading.
- Implement toolbox reloading - an API and quickly thrown together GUI. (I will drop the GUI if anyone complains it isn't so important.)
- Add monitor tag on toolbox root elements to force Galaxy to watch the toolbox for changes and reload as a result.
- Refactoring and test improvements for existing tool monitoring code.
- Implement toolbox shutdown process to ensure different watchers don't compete with each other.
- Set the default tool conf to be monitored.*
- Implement a generic parsing module to decouple toolbox creation from XML parsing.
- Implement a YAML toolbox conf parser that allows toolbox confs to be json/yaml. (JSON would be much better for an external automated component that would automatically install and configure tools from a theoretical new tool registry). 
- Enhancements to tool parsing and tool directory loading for use by other projects via galaxy-lib. (Warning: This includes some limited logic for parsing and reasoning about CWL tools but ABSOLUTELY none of my runtime plumbing. I swear my strategy is boiling a frog not attempting to sneak a giant horse in.)

\* Shed confs should never be monitored at this point, it would likely break everything. I talked with @davebx and we have agreed as a general design goal the tool shed 1.0 should move toward gutting all the code related to modifying tool confs in memory since much of it doesn't work well and it is inherently single process - and simply modify the shed_tool_conf and let Galaxy monitor the file for changes.

jmchilton committed:
Small improvements to load_directory for downstream extension.

jmchilton committed:
Add CWL suppoort to tool checking and parsing.

To be clear, this provides none of the infrastructure for running the tools. This is just so that planemo and future tool registry efforts can have a unified interface for accessing and reasoning about tools using the same methods Galaxy does.

jmchilton committed:
Refactor loader_directory to support beta formats in other ways.

jmchilton committed:
Infrastructure for parsing CWL tools.

jmchilton committed:
Introduce ToolBox parser.

Abstract simpler ToolBox interactions to separate XML representation from handling. Doesn't cover tool shed operations since the tool shed explicitly writes XML.

jmchilton committed:
Allow loading tool confs written in YAML and JSON.

Because lets say someone was writing a programatic way to manage elements in a tool conf, that hypothetical person probably shouldn't be writing XML.

jmchilton committed:
Allow toolbox reloading via app, UI, and API.

Introduce an app ToolCache to help ensure this operation remains speedy.

jmchilton committed:
Implement toolbox monitoring.

 - Add monitor tag on toolbox root elements to force Galaxy to watch the toolbox for changes.
 - Refactoring and test improvements for existing tool monitoring code.
 - Implement toolbox shutdown process to ensure different watchers don't compete with each other.
 - Set the default tool conf to be monitored.

jmchilton committed:
Revise tool caching for tool shed tool loading...

The tool shed appears to load the same tool in the toolbox multiple times with different settings. Get around this by not using the cache when loading tools from a few tool shed components later in this process.

jmchilton commented:
@galaxybot test this

jmchilton committed:
Refactor package_tool for better separation of concerns.

The tool should know how to package itself, it shouldn't be the responsiblity of the toolbox. Update the code to reflect this and use more pythonic exception handling.

jmchilton commented:
I think this is ready for merge, the API tests all passed on the last run through. I had this retested because three tool shed tests failed previously, but these three tests frequently fail and I believe  https://github.com/galaxyproject/galaxy/pull/1401 should address that problem.

martenson commented:
I would like to review this better if you are willing to give me some more
time. I can merge if otherwise.
M.

On Sat, Dec 26, 2015, 12:44 John Chilton notifications@github.com wrote:

> I think this is ready for merge, the API tests all passed on the last run
> through. I had this retested because three tool shed tests failed
> previously, but these three tests frequently fail and I believe #1401
> https://github.com/galaxyproject/galaxy/pull/1401 should address that
> problem.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/galaxyproject/galaxy/pull/1398#issuecomment-167348871
> .

jmchilton commented:
@martenson - Thanks! Take your time, that comment was only meant to prevent people from not merging because of the failing tests.

On the other hand... if you are offering to merge a PR without serious review https://github.com/galaxyproject/galaxy/pull/1137 has been open for over a month and been conflicted several times. :wink:

jmchilton commented:
Rebased and updated:
- b586204 - Has been rebased with a fix so that now Galaxy sleeps in monitor loop and doesn't make my laptop so hot.
- c46e2f9 - Contains a toolbox refactoring toward better separation of concerns between tools and toolboxes.
- e218445 - Contains a less obviously correct refactoring of toolbox stuff that I make the case for in the commit message.

martenson commented on templates/admin/reload_tool.mako:
```diff
@@ -14,6 +14,17 @@ $().ready(function() {
+            url: "${h.url_for(controller="/api/configuration", action="toolbox")}",
+            type: 'PUT',
+            success: function(result) {
+                alert("Toolbox reloaded.");
```
Can we please do alert when this fails instead? Or maybe just a console.log (in that case it can log both situations)

martenson commented on config/tool_conf.xml.sample:
```diff
@@ -1,5 +1,5 @@
 <?xml version='1.0' encoding='utf-8'?>
-<toolbox>
+<toolbox monitor="true">
```
My tool_conf.xml.sample was unchanged when I pulled these changes, some gitignore (that has `config/*`) thing?

jmchilton replied:
Yeah - I don't know how to fix this - any ideas?

bgruening replied:
I have this change if I pull in your changes: `<toolbox monitor="true">`
I don't think it has something do to with `config/*` as all the other config files also get's updated.

martenson replied:
I guess I have some local gitignore logic, or can you replicate @jmchilton?

jmchilton replied:
Actually it does work for me, I think I thought I replicated it but was looking at tool_conf.xml instead of the sample.

martenson commented:
When triggered this is probably trying to access files that do not have to exist?

```
galaxy.tools.toolbox.base DEBUG 2015-12-28 13:28:28,111 Loaded tool id: toolshed.g2.bx.psu.edu/repos/devteam/fastq_quality_boxplot/cshl_fastq_quality_boxplot/1.0.0, version: 1.0.0 into tool panel....
close failed in file object destructor:
IOError: [Errno 9] Bad file descriptor
galaxy.tools.deps DEBUG 2015-12-28 13:28:28,116 Unable to find config file './dependency_resolvers_conf.xml'
galaxy.tools.search DEBUG 2015-12-28 13:28:28,116 Starting to build toolbox index.
galaxy.tools.search DEBUG 2015-12-28 13:28:28,786 Toolbox index finished.
```

martenson commented:
note: comment out a tool; reload; remove comment; reload = tool will appear again but will be in different position (will be first in its section)

jmchilton committed:
Remove alert from toolbox reload.

jmchilton commented:
@martenson 

About:

```
galaxy.tools.toolbox.base DEBUG 2015-12-28 13:28:28,111 Loaded tool id: toolshed.g2.bx.psu.edu/repos/devteam/fastq_quality_boxplot/cshl_fastq_quality_boxplot/1.0.0, version: 1.0.0 into tool panel....
close failed in file object destructor:
IOError: [Errno 9] Bad file descriptor
galaxy.tools.deps DEBUG 2015-12-28 13:28:28,116 Unable to find config file './dependency_resolvers_conf.xml'
galaxy.tools.search DEBUG 2015-12-28 13:28:28,116 Starting to build toolbox index.
galaxy.tools.search DEBUG 2015-12-28 13:28:28,786 Toolbox index finished.
```

... does this occur when you start up Galaxy also? At first glance, this seems to be more a state conflict between the file system shed tool conf for your instance then related to this PR.

jmchilton commented:
@martenson "note: comment out a tool; reload; remove comment; reload = tool will appear again but will be in different position (will be first in its section)"

I cannot replicate this behavior with a normal tool conf. Is this a shed tool conf with multiple versions of that tool?

Update: Regardless, I imagine this could easily be a bug, but if it is a bug it is very likely a bug with the existing integrated tool configuration code and not with this pull request.

jmchilton commented:
@martenson Removed the alert, I don't want to be bike shed-ed on a terrible UI on a page we don't actually want though, so I stand by my threat to remove the GUI for this all together. This PR is much more important that link and that whole page, heck the whole admin section needs to be redone at some point soon.

jmchilton committed:
Fix shutdown of tool watcher if never activated.

Caused a harmless, but distracting stacktrace to be printed on toolbox reload as pointed out by @bgruening https://github.com/galaxyproject/galaxy/pull/1398#issuecomment-168499678.

bgruening commented:
![bildschirmfoto vom 2016-01-03 14-30-49](https://cloud.githubusercontent.com/assets/469983/12078962/cca1018a-b226-11e5-943d-a8cbff699ca1.png) ![bildschirmfoto vom 2016-01-03 14-27-52](https://cloud.githubusercontent.com/assets/469983/12078961/cc9dcf4c-b226-11e5-93ad-48abd80d556e.png)

@jmchilton I can the see same strange behaviour. See the above screenshots and have a look at the `merge columns` tool. But I'm pretty sure this is an old bug. I have seen this a few times before. It gets more strange if you include shed_tools as well.

bgruening commented:
If I use the UI (which I find useful) I get this error:

```
Traceback (most recent call last):
  File "/home/bag/projects/code/galaxy/lib/galaxy/web/framework/decorators.py", line 258, in decorator
    rval = func( self, trans, *args, **kwargs)
  File "/home/bag/projects/code/galaxy/lib/galaxy/web/framework/decorators.py", line 90, in decorator
    return func( self, trans, *args, **kwargs )
  File "/home/bag/projects/code/galaxy/lib/galaxy/webapps/galaxy/api/configuration.py", line 92, in reload_toolbox
    self.app.reload_toolbox()
  File "/home/bag/projects/code/galaxy/lib/galaxy/config.py", line 761, in reload_toolbox
    old_toolbox.shutdown()
  File "/home/bag/projects/code/galaxy/lib/galaxy/tools/toolbox/base.py", line 927, in shutdown
    raise exception
RuntimeError: cannot join thread before it is started

```

But it is working, so the ToolBox is reloaded. One additional feature would be to not load a tool conf if we have an XML-parse error in it. In this case nothing should happen but Galaxy should tell us with a big error message.

bgruening commented:
Since this only works for non-TS tools this should be added to the UI description. I guess many people will expect all tools to be reloaded. That said not sure we should hide that feature until the shed_tools are also working.

```
Enhancements to tool parsing and tool directory loading for use by other projects via galaxy-lib. (Warning: This includes some limited logic for parsing and reasoning about CWL tools but ABSOLUTELY none of my runtime plumbing. I swear my strategy is boiling a frog not attempting to sneak a giant horse in.)
```

Same as my last comments for your other PR. I'm not convinced that we should add this stuff, yet. 
I like the ToolBox enhancements in general and it would be awesome to have this for TS tools as well.

Thanks @jmchilton!

martenson commented:
@bgruening What do you mean that it works only for non-TS tools? It works for everything in `shed_tools_conf.xml` for me.

Also, this is very useful especially for managing instances like Main which I have been doing for a week in December and this would have prevented many restarts.

Also I am not getting `cannot join thread before it is started` and the error I have reported is not related to this PR.

I will merge this later today if nobody opposes it.

bgruening commented:
@martenson I have not tested shed_confs as John wrote `* Shed confs should never be monitored at this point, it would likely break everything.`
I will test this as well then.

Any idea what can cause the problem: `cannot join thread before it is started`

martenson commented:
@bgruening I think what @jmchilton meant is 'they are not monitored' in a sense they would reload automatically. So we have to click a link instead.

And no, I did not see that error before and don't know what causes it :(

jmchilton commented:
I can replicate the stacktrace pointed out by @bgruening, I'll try to fix it now. If the PR gets merged in the mean time I'll open a new one with a fix obviously.

jmchilton commented:
Fixed stack trace with d48fe04bb4dab855d4ab02564206360b4d1c6d69.

jmchilton commented:
@bgruening 

> Since this only works for non-TS tools this should be added to the UI description. I guess many people will expect all tools to be reloaded. That said not sure we should hide that feature until the shed_tools are also working.

Adding monitor to the conf should not be done with shed conf since that file will be modified by the tool shed code as it also modifies the internal in-memory representation of these things. The link in the UI should work for all confs however - including shed ones, it just reloads the whole toolbox.

jmchilton commented:
Rebased due to conflict introduced with merging #1409.

martenson commented:
Thanks @jmchilton !

jmchilton commented:
No - thank you for the detailed review and merge @martenson! Also thanks @bgruening.
